src/pubsub/ua_pubsub_reader.c

```diff
@@ -1449,17 +1536,75 @@ UA_Server_DataSetReader_process(UA_Server *server, UA_DataSetReader *dataSetRead
             }
         }
     }
+
+    UA_LOG_DEBUG(&server->config.logger, UA_LOGCATEGORY_SERVER, "DataSetReader '%.*s': received a network message", 
+        (int) dataSetReader->config.name.length, dataSetReader->config.name.data);
+#ifdef UA_ENABLE_PUBSUB_MONITORING
+    /* if previous reader state was error (because we haven't received messages and ran into timeout) we should set the state back to operational */
+    if (dataSetReader->state != UA_PUBSUBSTATE_OPERATIONAL) {
+        UA_DataSetReader_setPubSubState(server, UA_PUBSUBSTATE_OPERATIONAL, dataSetReader);
+    }
+    if (dataSetReader->msgRcvTimeoutTimerRunning == UA_TRUE) {
+        /* stop message receive timeout timer */
+        if (server->config.pubsubConfiguration->monitoringInterface.stopMonitoring(server, dataSetReader->identifier, 
+            UA_PUBSUB_COMPONENT_DATASETREADER, eMessageReceiveTimeout, dataSetReader) == UA_STATUSCODE_GOOD) {
+            dataSetReader->msgRcvTimeoutTimerRunning = UA_FALSE;
+        } else {
+            UA_LOG_ERROR(&server->config.logger, UA_LOGCATEGORY_SERVER, "DataSetReader '%.*s': stop receive timeout timer failed",
+                (int) dataSetReader->config.name.length, dataSetReader->config.name.data);
+            UA_DataSetReader_setPubSubState(server, UA_PUBSUBSTATE_ERROR, dataSetReader);
+        }
+    }
+    /* start message receive timeout timer */
+    if (server->config.pubsubConfiguration->monitoringInterface.startMonitoring(server, dataSetReader->identifier, 
+            UA_PUBSUB_COMPONENT_DATASETREADER, eMessageReceiveTimeout, dataSetReader) == UA_STATUSCODE_GOOD) {
+        UA_LOG_DEBUG(&server->config.logger, UA_LOGCATEGORY_SERVER, "Info: DataSetReader '%.*s': start receive timeout timer",
+            (int) dataSetReader->config.name.length, dataSetReader->config.name.data);
+        dataSetReader->msgRcvTimeoutTimerRunning = UA_TRUE;
+    } else {
+        UA_LOG_ERROR(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                        "Starting Message Receive Timeout timer failed.");
+        UA_DataSetReader_setPubSubState(server, UA_PUBSUBSTATE_ERROR, dataSetReader);
```

Comment:
Just for interest: Did you make some profiling? This code path is called very often.